deepspeech.go

```diff
@@ -63,9 +63,50 @@ type sliceHeader struct {
 // buffer     A 16-bit, mono raw audio signal at the appropriate sample rate.
 // bufferSize The number of samples in the audio signal.
 // sampleRate The sample-rate of the audio signal.
-// TODO Make sure the C string is cleaned properly
 func (m *Model) SpeechToText(buffer []int16, bufferSize, sampleRate uint) string {
-	return C.GoString(C.STT(m.w, (*C.short)(unsafe.Pointer((*sliceHeader)(unsafe.Pointer(&buffer)).Data)), C.uint(bufferSize), C.uint(sampleRate)))
+	str := C.STT(m.w, (*C.short)(unsafe.Pointer((*sliceHeader)(unsafe.Pointer(&buffer)).Data)), C.uint(bufferSize), C.uint(sampleRate))
+	retval := C.GoString(str)
+	defer C.FreeString(str)
+	return retval
+}
+
+type MetadataItem C.struct_MetadataItem
+
+func (mi *MetadataItem) Character() string {
+	return C.GoString(C.MetadataItem_GetCharacter((*C.MetadataItem)(unsafe.Pointer(mi))))
+}
+
+func (mi *MetadataItem) Timestep() int {
+	return int(C.MetadataItem_GetTimestep((*C.MetadataItem)(unsafe.Pointer(mi))))
+}
+
+func (mi *MetadataItem) StartTime() float32 {
+	return float32(C.MetadataItem_GetStartTime((*C.MetadataItem)(unsafe.Pointer(mi))))
+}
+
+// Metadata represents a DeepSpeech metadata output
+type Metadata C.struct_Metadata
+
+func (m *Metadata) NumItems() int32 {
+	return int32(C.Metadata_GetNumItems((*C.Metadata)(unsafe.Pointer(m))))
+}
+
+func (m *Metadata) Probability() float64 {
+	return float64(C.Metadata_GetProbability((*C.Metadata)(unsafe.Pointer(m))))
+}
+
+func (m *Metadata) Items() []MetadataItem {
+	num_items := int32(C.Metadata_GetNumItems((*C.Metadata)(unsafe.Pointer(m))))
+	allItems := C.Metadata_GetItems((*C.Metadata)(unsafe.Pointer(m)))
+	return (*[1 << 30]MetadataItem)(unsafe.Pointer(allItems))[:num_items:num_items]
+}
+
+// SpeechToTextWithMetadata uses the DeepSpeech model to perform Speech-To-Text.
+// buffer     A 16-bit, mono raw audio signal at the appropriate sample rate.
+// bufferSize The number of samples in the audio signal.
+// sampleRate The sample-rate of the audio signal.
+func (m *Model) SpeechToTextWithMetadata(buffer []int16, bufferSize, sampleRate uint) *Metadata {
+	return (*Metadata)(unsafe.Pointer(C.STTWithMetadata(m.w, (*C.short)(unsafe.Pointer((*sliceHeader)(unsafe.Pointer(&buffer)).Data)), C.uint(bufferSize), C.uint(sampleRate))))
```

Comment:
I'm not sure how to deal with `DS_FreeMetadata() / C.FreeMetadata()` when doing things like that. I've removed the `MetadataWrapper` that felt useless, should I keep it only for that purpose ? Or is it okay to expect user the call `defer C.FreeMetadata()` after `SpeechToTextWithMetadata` ?

Reply:
Usually in GO, when there's a logic that should be executed by the user to clean the variable, we use the `Closer` interface that implements `Close() error`.

Therefore you should add such a method to `Metadata` that executes `C.FreeMetadata()`. And in the client, you should add `defer m.Close()` after `SpeechToTextWithMetadata`.

Reply:
Thanks, I've added a `Close()` as suggested.